internal/manager/run.go

```diff
@@ -70,6 +70,13 @@ func Run(ctx context.Context, c *Config, diagnostic util.ConfigDumpDiagnostic, d
 	if err != nil {
 		return fmt.Errorf("could not validate Kong admin root(s) configuration: %w", err)
 	}
+
+	// gateway discovery is only supported in db-less mode in its initial release:
+	// https://github.com/Kong/kubernetes-ingress-controller/issues/3401
+	if c.KongAdminSvc.IsPresent() && !((dbMode == "off") || (dbMode == "")) {
+		return fmt.Errorf("gateway discovery is only supported in dbless mode, not in db %s", dbMode)
+	}
```

Comment:
I know it's a very simple condition, but I'd feel safe if we had a separate function that we could cover with unit tests. This could even be the already existing `kongconfig.ValidateRoots` (we could pass `c.KongAdminSvc` to it and cover all the cases with unit tests).

My rationale for such a maneuver is that it's better to keep such business logic somewhere else than setup code + it makes it unit testable.

Reply:
extracted to a function `c.ValidateGatewayDiscovery()` and added unit tests.